### ISSUE TITLE
[1.20 regression] "Unsupported operand types for |" with recursive type definition

### ISSUE THREAD

**Issue by latentvoid** [python/mypy#21141]
**Bug Report**

When merging dicts with a value type that is union containing a recursive type (e.g. `A = dict[str, A | int]`), mypy incorrectly detects an "Unsupported operand types" error. This worked in version 1.19.1.

**To Reproduce**

```python
type A = dict[str, A | int]		# can use arbitrary types instead of `str` and `int`

d1: A
d2: A
d3: A = {**d1, **d2}		# works
d4: A = d1 | d2		# equivalent to line above but fails
d5 = d1 | d2
d6: A = d5		# works (similar behavior to what was reported in https://github.com/python/mypy/issues/18236)

type B = dict[str, B]

d7: B
d8: B
d9: B = d7 | d8		# no union in type def works

type C = dict[str, float | int]

d10: C
d11: C
d12: C = d10 | d11		# no recursive type def works
```

Playground URL: https://mypy-play.net/?mypy=master&python=3.12&gist=1c39fc367614e2a0af2bd4d604c25aeb

The mypy playground seems to be bugged currently. Selecting version "mypy 1.20.0" results in HTTP 500 and "mypy latest (1.20.0)" does not reproduce the bug even though it should. "mypy master branch" does reproduce the bug.

**Expected Behavior**

None of the examples above should produce an error.

**Actual Behavior**

Mypy produces `error: Unsupported operand types for | ("dict[str, A | int]" and "dict[str, A | int]")  [operator]`

**Your Environment**

- Mypy version used: 1.20.0
- Mypy command-line flags: none
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: 3.12.12, 3.12.13, 3.14.3

**Comment by hauntsaninja** [comment]
Thanks, this bisects to a change in typeshed #21021

**Comment by ilevkivskyi** [comment]
I have a wild guess this may be caused by a change in typeshed https://github.com/python/typeshed/pull/14284. I will check later and if yes, I propose to revert that.

**Comment by hauntsaninja** [comment]
Probably https://github.com/python/typeshed/pull/14284

**Comment by hauntsaninja** [comment]
Oh I'm too slow :-)

**Comment by flying-sheep** [comment]
this isn’t even constrained to recursive type definitions, although I have no clue how to reproduce this in the playground:

https://github.com/scverse/session-info2/blob/a99a8777d31d6a081aefa527ff9b6a3d48b0c476/tests/test_synthetic.py#L97

https://results.pre-commit.ci/run/github/734840558/1775498155.jbv0e9lgRGiQ0qRpSFUYeg

```
  tests/test_synthetic.py:97: error: Unsupported operand types for | ("dict[str, list[str]]" and "dict[str, list[str]]")  [operator]
```

**Comment by ilevkivskyi** [comment]
Here the repro distilled from your case:
```python
from typing import Mapping, Sequence

d: dict[str, list[str]]
m: Mapping[str, Sequence[str]] = d | d
```
Btw @hauntsaninja your PR doesn't fix this case. I really think we should proceed with the revert.

**Comment by AA-Turner** [comment]
For posterity, another example -- this seems to be fixed by the revert:

```python
from __future__ import annotations

from typing import Any


class Item:
    a: str
    b: str | None
    c: str
    d: bytes
    e: int
    f: bytes
    z: bool

    def __init__(self, a: str, b: str | None, c: str, d: bytes, e: int, f: bytes, z: bool) -> None:
        self.a = a
        self.b = b
        self.c = c
        self.d = d
        self.e = e
        self.f = f
        self.z = z


def new_item(d: dict[str, Any]) -> Item:
    return Item(**(d | {"z": False}))
```

```shell
$ uvx mypy==1.20
bug.py:26: error: Argument 1 to "Item" has incompatible type "**dict[str, Any | bool]"; expected "str"  [arg-type]
bug.py:26: error: Argument 1 to "Item" has incompatible type "**dict[str, Any | bool]"; expected "str | None"  [arg-type]
bug.py:26: error: Argument 1 to "Item" has incompatible type "**dict[str, Any | bool]"; expected "bytes"  [arg-type]
Found 3 errors in 1 file (checked 1 source file)
```

A
